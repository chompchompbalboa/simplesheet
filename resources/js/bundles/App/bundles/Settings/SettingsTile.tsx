//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React, { useState } from 'react'
import styled from 'styled-components'

//-----------------------------------------------------------------------------
// Component
//-----------------------------------------------------------------------------
const SettingsTile = ({
  children,
  header
}: SettingsTileProps) => {

  const [ isItemsVisible, setIsItemsVisible ] = useState(true)

  return (
    <Container>
      <Header
        data-testid="settingsTileHeader" 
        onClick={() => setIsItemsVisible(!isItemsVisible)}>
        {header}
      </Header>
      <Items 
        data-testid="settingsTileItems"
        isItemsVisible={isItemsVisible}>
        {children}
      </Items>
    </Container>
  )
}

//-----------------------------------------------------------------------------
// Props
//-----------------------------------------------------------------------------
export type SettingsTileProps = {
  children?: any
  header: string
}

//-----------------------------------------------------------------------------
// Styled Components
//-----------------------------------------------------------------------------
const Container = styled.div`
`

const Header = styled.h3`
  margin-left: 0.25rem;
  cursor: pointer;
  width: 100%;
  &:hover {
    text-decoration: underline;
  }
`

const Items = styled.div`
  display: ${ ({ isItemsVisible }: ItemsProps) => isItemsVisible ? 'block' : 'none'};
`
type ItemsProps = {
  isItemsVisible: boolean
}

export default SettingsTile
