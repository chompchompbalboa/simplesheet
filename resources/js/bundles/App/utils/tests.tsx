//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunkMiddleware from 'redux-thunk'
import { render } from '@testing-library/react'

import { appReducer } from '@app/state'

//-----------------------------------------------------------------------------
// With Redux
//-----------------------------------------------------------------------------
// @ts-ignore
const WithRedux = ({ children }) => {
  const store = createStore(appReducer, applyMiddleware(thunkMiddleware))
  return (
    <ReduxProvider store={store}>
      { children }
    </ReduxProvider>
  )
}

//-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------
export * from '@testing-library/react'
// @ts-ignore
export const renderWithRedux = (ui, options?) => render(ui, { wrapper: WithRedux, ...options })
