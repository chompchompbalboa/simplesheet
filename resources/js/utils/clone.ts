const clone = (itemToClone: any): any => {
	return JSON.parse(JSON.stringify(itemToClone))
}

export default clone
