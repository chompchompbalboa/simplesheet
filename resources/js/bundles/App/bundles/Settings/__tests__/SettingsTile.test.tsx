//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React from 'react'
import 'jest-styled-components'
import { cleanup, fireEvent, render } from '@app/utils/tests'
import SettingsTile, { SettingsTileProps } from '../SettingsTile'

//-----------------------------------------------------------------------------
// Setup
//-----------------------------------------------------------------------------
const props: SettingsTileProps = {
  header: "Header"
}

afterEach(cleanup)

//-----------------------------------------------------------------------------
// Tests
//-----------------------------------------------------------------------------
test("Renders correctly", async () => {
  const { container } = render(<SettingsTile {...props}/>)
  expect(container).toMatchSnapshot()
})

test("Opens and closes when clicking the link", async () => {
  const { getByTestId } = render(<SettingsTile  {...props}/>)
  const settingsTileHeader = getByTestId('settingsTileHeader')
  const settingsTileItems = getByTestId('settingsTileItems')
  expect(settingsTileItems).toHaveStyleRule('display', 'none')
  fireEvent.click(settingsTileHeader)
  expect(settingsTileItems).toHaveStyleRule('display', 'block')
  fireEvent.click(settingsTileHeader)
  expect(settingsTileItems).toHaveStyleRule('display', 'none')
})