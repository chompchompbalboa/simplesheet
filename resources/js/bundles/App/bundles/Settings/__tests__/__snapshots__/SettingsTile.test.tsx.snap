// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
.c0 {
  cursor: pointer;
  width: 100%;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 {
  display: none;
}

<div>
  <div
    class=""
  >
    <h3
      class="c0"
      data-testid="settingsTileHeader"
    >
      Header
    </h3>
    <div
      class="c1"
      data-testid="settingsTileItems"
    />
  </div>
</div>
`;
